azure-kusto-data/source/connectionBuilder.ts

```diff
@@ -2,87 +2,111 @@
-        mappedTo: "application Certificate PrivateKey",
-        validNames: ["application Certificate PrivateKey"]
+        mappedTo: "Application Certificate PrivateKey",
+        validNames: ["Application Certificate PrivateKey"],
+        isSecret: true,
     },
     applicationCertificateThumbprint: {
-        propName: "applicationCertificateThumbprint",
         mappedTo: "Application Certificate Thumbprint",
-        validNames: ["application certificate thumbprint"]
+        validNames: ["application certificate thumbprint", "AppCert"],
     },
     applicationCertificateX5c: {
-        propName: "applicationCertificateX5c",
         mappedTo: "Application Certificate x5c",
-        validNames: ["application certificate x5c"]
+        validNames: ["application certificate x5c", "Application Certificate Send Public Certificate", "Application Certificate SendX5c", "SendX5c"],
     },
     authorityId: {
-        propName: "authorityId",
         mappedTo: "Authority Id",
-        validNames: ["authority id", "authorityid", "authority", "tenantid", "tenant", "tid"]
-    }
+        validNames: ["authority id", "authorityid", "authority", "tenantid", "tenant", "tid"],
+    },
 });
 
-const getPropName = (key: string): string => {
+const getPropName = (key: string): [string, MappingType] => {
     const _key = key.trim().toLowerCase();
 
     for (const keyword of Object.keys(KeywordMapping)) {
-        const k = KeywordMapping[keyword];
-        if (k.validNames.indexOf(_key) >= 0) {
-            return k.propName;
+        const k = KeywordMapping[keyword as KcsbMappedKeys];
+        if (!k) {
+            continue;
+        }
+        if (k.validNames.map(n => n.trim().toLowerCase()).indexOf(_key) >= 0) {
+            return [keyword, k];
         }
     }
-    throw new Error(key);
+    throw new Error("Failed to get prop: " + key);
 };
 
+
 export class KustoConnectionStringBuilder {
-    [prop: string]: string | boolean | ((info: DeviceCodeResponse) => void) |  undefined;
+    static readonly SecretReplacement = "****";
+    // tslint:disable-next-line:no-console
+    static defaultDeviceCallback: (response: DeviceCodeResponse) => void = (response) => console.log(response.message);
+
     dataSource?: string;
+    aadFederatedSecurity?: boolean;
     aadUserId?: string;
     password?: string;
     applicationClientId?: string;
+    msiClientId?: string;
     applicationKey?: string;
     applicationCertificatePrivateKey?: string;
     applicationCertificateThumbprint?: string;
-    authorityId?: string;
+    applicationCertificateX5c?: string;
+    authorityId: string = "common";
     deviceCodeCallback?: (response: DeviceCodeResponse) => void;
+    tokenProvider?: () => Promise<string>;
+    loginHint?: string;
+    timeoutMs?: number;
+    accessToken?: string;
+    isDeviceCode?: boolean;
+    isInteractiveLogin?: boolean;
+    isAzLoginIdentity?: boolean;
+    isManagedIdentity?: boolean;
```

Comment:
Can be replaced by enum?

Reply:
Just for these four, or have an enum member for each type (appkey, userpass, etc..)?
If the user doesn't provide the enum, is that an error?